src/dhcpmon/src/dhcp_device.c

```diff
@@ -414,6 +419,56 @@ int dhcp_device_init(dhcp_device_context_t **context,
     return rv;
 }
 
+/**
+ * @code dhcp_device_start_capture(context, snaplen, base, vlan_ip);
+ *
+ * @brief starts packet capture on this interface
+ */
+int dhcp_device_start_capture(dhcp_device_context_t *context,
+                              size_t snaplen,
+                              struct event_base *base,
+                              in_addr_t vlan_ip)
+{
+    int rv = -1;
+
+    do {
+        if (context == NULL) {
+            syslog(LOG_ALERT, "NULL interface context pointer'\n");
+            break;
+        }
+
+        if (snaplen < UDP_START_OFFSET + sizeof(struct udphdr) + DHCP_OPTIONS_HEADER_SIZE) {
+            syslog(LOG_ALERT, "dhcp_device_start_capture(%s): snap length is too low to capture DHCP options", context->intf);
+            break;
+        }
+
+        context->vlan_ip = vlan_ip;
+
+        context->buffer = (uint8_t *) malloc(snaplen);
+        if (context->buffer == NULL) {
+            syslog(LOG_ALERT, "malloc: failed to allocate memory for socket buffer '%s'\n", strerror(errno));
+            break;
+        }
+        context->snaplen = snaplen;
+
+        if (setsockopt(context->sock, SOL_SOCKET, SO_ATTACH_FILTER, &dhcp_sock_bfp, sizeof(dhcp_sock_bfp)) != 0) {
+            syslog(LOG_ALERT, "setsockopt: failed to attach filter with '%s'\n", strerror(errno));
+            break;
+        }
+
+        struct event *ev = event_new(base, context->sock, EV_READ | EV_PERSIST, read_callback, context);
+        if (ev == NULL) {
+            syslog(LOG_ALERT, "event_new: failed to allocate memory for libevent event '%s'\n", strerror(errno));
+            break;
+        }
+        event_add(ev, NULL);
+
+        rv = 0;
+    } while (0);
```

Comment:
What is the purpose of the `do { ... } while(0)` loop? It will always execute exactly once. Can we simply remove the loop construct?

Reply:
This is to avoid chevron style (or use of go to statement) coding since each failure in if clause would cause the block to abort.

Reply:
I see. In that case, you could instead simplify as follows (also taking advantage of different error codes here):

```c
int dhcp_device_start_capture(dhcp_device_context_t *context,
                              size_t snaplen,
                              struct event_base *base,
                              in_addr_t vlan_ip)
{
    if (context == NULL) {
        syslog(LOG_ALERT, "NULL interface context pointer'\n");
        return -1;
    }

    if (snaplen < UDP_START_OFFSET + sizeof(struct udphdr) + DHCP_OPTIONS_HEADER_SIZE) {
        syslog(LOG_ALERT, "dhcp_device_start_capture(%s): snap length is too low to capture DHCP options", context->intf);
        return -2;
    }

    context->vlan_ip = vlan_ip;

    context->buffer = (uint8_t *) malloc(snaplen);
    if (context->buffer == NULL) {
        syslog(LOG_ALERT, "malloc: failed to allocate memory for socket buffer '%s'\n", strerror(errno));
        return -3;
    }
    context->snaplen = snaplen;

    if (setsockopt(context->sock, SOL_SOCKET, SO_ATTACH_FILTER, &dhcp_sock_bfp, sizeof(dhcp_sock_bfp)) != 0) {
        syslog(LOG_ALERT, "setsockopt: failed to attach filter with '%s'\n", strerror(errno));
        return -4;
    }

    struct event *ev = event_new(base, context->sock, EV_READ | EV_PERSIST, read_callback, context);
    if (ev == NULL) {
        syslog(LOG_ALERT, "event_new: failed to allocate memory for libevent event '%s'\n", strerror(errno));
        return -5;
    }

    event_add(ev, NULL);

    return 0;
}
```

Reply:
It is very similar, however some static analysis and code standard deprecate having multiple return point. It make formal verification a bit difficult.:

https://spin.atomicobject.com/2011/07/26/in-defence-of-misra/

Reply:
Valid points, but as long as the logic is fairly linear and the return codes are unique, I think it's acceptable. I prefer it to using a loop construct when it's not necessary.

Reply:
Since there are other similar constructs throughout, this can be saved for a separate PR focused on refactoring.